### PR TITLE
Document RFC 1623: static lifetime elision.

### PR DESCRIPTION
**Pull request by chriskrycho**
This should be the last item required for stabilizing RFC 1623 (#35897).

### CODE REVIEW

File: src/doc/reference.md [hunk 1]
```diff
@@ -1269,15 +1269,14 @@ guaranteed to refer to the same memory address.
 
 Constant values must not have destructors, and otherwise permit most forms of
 data. Constants may refer to the address of other constants, in which case the
-address will have the `static` lifetime. The compiler is, however, still at
-liberty to translate the constant many times, so the address referred to may not
-be stable.
+address will have the `static` lifetime. (See below on [static lifetime elision](#static-lifetime-elision).) The compiler is, however, still at liberty to
```
Comment (frewsxcv) [inline src/doc/reference.md]: Can you wrap this line to match the indentation of the surrounding lines?

**Comment by chriskrycho** [comment]
Gah, yes. I’m annoyed I missed that.

**Comment by steveklabnik** [comment]
So, @rust-lang/docs , this raises a question. How do we make the distinction between stable and unstable stuff here?

**Comment by chriskrycho** [comment]
I think at some point, we are going to need to build infrastructure for versioning. That is going to be some work, but I do not see how we can handle these kinds of cases long-term without it. I can open an issue to that effect if you like. And maybe ping some of the Ember people who worked on their infrastructure that way.

**Comment by GuillaumeGomez** [comment]
It should have an unstable version. I don't have much better idea. :-/

**Comment by chriskrycho** [comment]
I just noted that all the code samples in reference.md get tested. Neat! I'll update the sample accordingly in the morning.

**Comment by frewsxcv** [comment]
I don't have any great ideas either, other than explicitly mentioning that it's unstable and adding in the `#[feature(whatever_feature_name)]` in the code example. At least we'll get a warning when it stabilizes.

**Comment by steveklabnik** [comment]
Putting this on this week's doc team agenda.

**Comment by llogiq** [comment]
We could change the feature implementation to my first commit, thus stabilizing the feature and removing the extra code adding the feature gate that makes it unstable in the first place. I presume this should make the doc tests pass.

**Comment by steveklabnik** [comment]
So, we talked about this at the docs meeting today.

What it really boils down to is this: it'd be a shame to block this based on figuring out the broader story here. However, once we know that the rest of the feature will be stable, we can just land this, then land the stabilization of the feature. @aturon [just put the feature into FCP](https://github.com/rust-lang/rust/issues/35897#issuecomment-267177618), so let's see how that goes before we merge.

And in the meantime, let's figure out the broader docs story. I'm going to open an issue on the RFCs repo about this.

**Comment by chriskrycho** [comment]
Related: I'll have time Friday or Saturday to update the not-the-feature-gate bits of the test that are broken here—there's one piece that needs to be tweaked independent of the stabilization/feature flag issue.

**Comment by chriskrycho** [comment]
Kept spacing this; came back to it via other issues. I'll update the non-feature-gate related parts tonight or tomorrow!

**Comment by llogiq** [comment]
@chriskrycho do you need any help?

**Comment by chriskrycho** [comment]
@llogiq gah. Slipped my mind. The only thing that needs tweaking is setting `ignore` on [this example](https://github.com/rust-lang/rust/pull/37928/files#diff-070c91ddb3b61fa408b94f592961734aR1364). It's re-using items from earlier in the reference, but they're not in scope for that specific example.

**Comment by chriskrycho** [comment]
I just pushed a fixed version of the commit which just includes all the required pieces. We'll see what Travis says…

**Comment by chriskrycho** [comment]
UGH. I pushed the rebase, but *not the fix*.

This is how my whole first day back at work has been. 🤦‍♂️

**Comment by steveklabnik** [comment]
@chriskrycho travis is still failing: https://travis-ci.org/rust-lang/rust/jobs/188704796#L1662

**Comment by chriskrycho** [comment]
🤔 Grr. I will follow up Wednesday evening. Family commitments till then.

**Comment by steveklabnik** [comment]
No worries!

**Review by llogiq — COMMENTED** [review]
Actually this is full elision with a `'static` default. This means that if the types contains `fn`s, the usual elision rules will apply, only choosing `'static` if no elision rule applies.

**Comment by chriskrycho** [comment]
@llogiq can you elaborate?

@steveklabnik I seem to have failed in pacifying [Travis](https://travis-ci.org/rust-lang/rust/jobs/196180998#L5822) and the feature gate. Can you take a gander and indicate what I did wrong?

**Comment by llogiq** [comment]
This means if you have a `static FOO : &Fn(&Bar) -> &Baz`, you actually have a `static FOO : &'static Fn<'a>(&'a Bar) -> &'a Baz` for some unnamed `'a`.

**Comment by est31** [comment]
> I seem to have failed in pacifying Travis and the feature gate. 

I think you can fix it by adding `# #![feature(...)]` to the doc test. It won't show up in the output then due to the leading `#`, while fixing the issue. See also [the book](https://doc.rust-lang.org/book/documentation.html#documentation-as-tests) and [the asm documentation](https://github.com/rust-lang/rust/tree/master/src/doc/book/inline-assembly.md), it uses this as well.

**Comment by chriskrycho** [comment]
@est31 ah, it's the extra `#`. Got it.

@llogiq do you have a suggested change to the write-up here?

File: src/doc/reference.md [hunk 2]
```diff
@@ -1355,6 +1356,32 @@ unsafe fn bump_levels_unsafe2() -> u32 {
...
 
+#### `'static` lifetime elision
+
+[Unstable] Both constant and static declarations of reference types have
+*implicit* `'static` lifetimes unless an explicit lifetime is specified. As
+such, the constant declarations involving `'static` above may be written
+without the lifetimes. Returning to our previous example:
+
+```rust
+#[feature(static_in_const)]
```
Comment (clarcharr) [inline src/doc/reference.md]: this should be `#![feature(static_in_const)]`; you forgot the exclamation point.

Comment (chriskrycho) [inline src/doc/reference.md]: YOU ARE MY HERO

**Comment by llogiq** [comment]
I think I *could* improve it, but I need to find the time first.

File: src/doc/reference.md [hunk 3]
```diff
@@ -1271,15 +1271,16 @@ guaranteed to refer to the same memory address.
 
 Constant values must not have destructors, and otherwise permit most forms of
 data. Constants may refer to the address of other constants, in which case the
-address will have the `static` lifetime. The compiler is, however, still at
+address will have the `static` lifetime. (See below on [static lifetime
```
Comment (llogiq) [inline src/doc/reference.md]: How about "will have elided lifetimes where applicable, otherwise – in most cases – defaulting to the `'static` lifetime.

**Review by llogiq — CHANGES_REQUESTED** [review]
Also after the example, add a paragraph to explain the elision:

In case the static or constant items contain function references or closures, and those work with references, the compiler will try the usual elision rules (link to elision). Failing that, it will default the lifetimes to `'static`. This is best explained by an example:

```rust
const FUN : fn(&str) -> &str = ..
// per rule #1, this is fn<'a>(&'a str) -> &'a str

const FUNNY : Fn(&Foo, &Bar, &Baz) -> usize = ..
// per rule #2, this is Fn<'a, 'b, 'c>(&'a Foo, &'b Bar, &'c Baz) -> usize

const UNFUNNY : Fn(&Foo, &Bar) -> &Baz = ..
// neither rule applies, so all `&`s are `'static`
```

**Comment by est31** [comment]
Now that beta branched, I would like to ask for beta backport nomination, so that already 1.16 would have it stabilized.

**Comment by llogiq** [comment]
Good idea.

**Comment by steveklabnik** [comment]
@chriskrycho ping! any interest in keeping this going? this is still blocking stabilization.

@est31 I would be strongly against this; stabilization backports are incredibly rare.

**Comment by chriskrycho** [comment]
@steveklabnik yeah, should be able to allocate an hour to it tomorrow. I'll pull in @llogiq's comments, as well as fix the build issue, then.

**Comment by llogiq** [comment]
We may also want to show embedded lifetimes, e.g. `Fn(&'a Foo) -> Bar<'a>` can be shortened to `Fn(&Foo) -> Bar`.

**Comment by chriskrycho** [comment]
@llogiq regarding your last comment: did that change with this RFC/implementation? If not, I'd prefer to note it as something to add in a separate commit/PR. I think we actually need a dedicated section for lifetime elision in general; right now it appears to be documented in detail only in the nomicon.

**Comment by chriskrycho** [comment]
Updated with everything other than the *embedded lifetimes* scenario, and it passed `rustup run nightly rustdoc --test ./doc/reference.md` so we should finally be good to go. 🤞

**Comment by steveklabnik** [comment]
Travis passed!

@bors: r+ rollup

thanks a ton @chriskrycho

**Comment by chriskrycho** [comment]
YESSSSSSSSSSS. FINALLLYYYYYYYY.

Sorry for the many delays on this, everyone! Glad it's finally in. So it'll land in, what, 1.17?

**Comment by est31** [comment]
@chriskrycho 

> So it'll land in, what, 1.17?

yes.

**Comment by alexcrichton** [comment]
Removing beta nomination as sentiment seems to be to not backport (also see #35897)

**Pull request merged**